File: apps/workers/src/middleware/origin-guard.ts
import { createLogger } from "@lumen/logger";
import type Elysia from "elysia";
import { env } from "../env";

const logger = createLogger({ name: "workers:origin-guard" });

export const DEFAULT_ALLOWED_ORIGINS = [
  "http://localhost:3000",
  "http://127.0.0.1:3000",
  "http://localhost:1420",
  "tauri://localhost",
  "http://tauri.localhost",
  "https://tauri.localhost",
];

const LOOPBACK_HOSTS = new Set(["localhost", "127.0.0.1", "0.0.0.0", "::1"]);

// Check whether an incoming origin or referer is in the allowed list or is a valid loopback alias
export function isAllowedOrigin(
  originOrReferer: string | null | undefined,
  configuredOrigins?: string[]
): boolean {
  if (!originOrReferer) {
    return false;
  }

  const allowedList = [
    ...(configuredOrigins ?? env.ALLOWED_ORIGINS ?? []),
    ...DEFAULT_ALLOWED_ORIGINS,
  ];

  let rawOrigin = originOrReferer;
  try {
    if (
      originOrReferer.startsWith("http://") ||
      originOrReferer.startsWith("https://")
    ) {
      rawOrigin = new URL(originOrReferer).origin;
    }
  } catch {
    // If not parseable as URL, fall back to raw string comparison
  }

  for (const allowed of allowedList) {
    if (rawOrigin === allowed) {
      return true;
    }

    try {
      if (
        (rawOrigin.startsWith("http://") || rawOrigin.startsWith("https://")) &&
        (allowed.startsWith("http://") || allowed.startsWith("https://"))
      ) {
        const allowedUrl = new URL(allowed);
        const incomingUrl = new URL(rawOrigin);

        if (
          allowedUrl.protocol === incomingUrl.protocol &&
          allowedUrl.port === incomingUrl.port &&
          LOOPBACK_HOSTS.has(allowedUrl.hostname) &&
          LOOPBACK_HOSTS.has(incomingUrl.hostname)
        ) {
          return true;
        }
      }
    } catch {
      // Ignore non-standard URL schemes
    }
  }

  return false;
}

export const originGuard = (app: Elysia) => {
  return app.onRequest(({ request, set }) => {
    // 1. Allow CORS preflight requests
    if (request.method === "OPTIONS") {
      return;
    }

    const url = new URL(request.url);
    const path = url.pathname;

    // 2. Allow JWKS endpoint (RFC 7517 public cryptographic keys) and OAuth callback redirects
    if (path === "/api/auth/jwks" || path.startsWith("/api/auth/callback/")) {
      return;
    }

    // 3. Allow internal service communication via internal API key
    const internalKey = request.headers.get("x-internal-key");
    const authHeader = request.headers.get("authorization");
    const configuredKey = env.INTERNAL_API_KEY || process.env.INTERNAL_API_KEY;
    if (
      configuredKey &&
      (internalKey === configuredKey ||
        authHeader === `Bearer ${configuredKey}`)
    ) {
      return;
    }

    // 4. Allow test harness / e2e bypass in non-production
    const e2eBypass = request.headers.get("x-e2e-bypass");
    if (
      env.NODE_ENV !== "production" &&
      (e2eBypass === "1" || e2eBypass === "true")
    ) {
      return;
    }

    // 5. Block direct browser navigation (typing URL in address bar or document embedding)
    const secFetchDest = request.headers.get("sec-fetch-dest");
    const secFetchMode = request.headers.get("sec-fetch-mode");
    if (secFetchDest === "document" || secFetchMode === "navigate") {
      logger.warn("Blocked direct browser navigation to workers API", {
        path,
        secFetchDest,
        secFetchMode,
      });
      set.status = 403;
      return {
        error: "Forbidden",
        message: "Direct browser access is blocked",
      };
    }

    // 6. In unit test environment, allow requests if no explicit origin header was passed
    if (env.NODE_ENV === "test" && !request.headers.has("origin")) {
      return;
    }

    // 7. Check Origin or Referer header
    const origin = request.headers.get("origin");
    const referer = request.headers.get("referer");

    if (origin && isAllowedOrigin(origin)) {
      return;
    }

    if (referer && isAllowedOrigin(referer)) {
      return;
    }

    logger.warn("Blocked request from untrusted origin", {
      path,
      origin: origin ?? "none",
      referer: referer ?? "none",
      method: request.method,
    });

    set.status = 403;
    return {
      error: "Forbidden",
      message: "Direct access or untrusted origin is blocked",
    };
  });
};

File: apps/presence/lib/presence_web/plugs/origin_guard.ex
defmodule PresenceWeb.Plugs.OriginGuard do
  @moduledoc """
  Blocks direct browser access and requests from untrusted origins for the presence service.
  """

  import Plug.Conn

  require Logger

  @default_allowed_origins [
    "http://localhost:3000",
    "http://127.0.0.1:3000",
    "http://localhost:1420",
    "tauri://localhost",
    "http://tauri.localhost",
    "https://tauri.localhost"
  ]

  @loopback_hosts ["localhost", "127.0.0.1", "0.0.0.0", "::1"]

  def init(opts), do: opts

  def call(conn, _opts) do
    # Allow OPTIONS requests (CORS preflight)
    if conn.method == "OPTIONS" do
      conn
    else
      check_access(conn)
    end
  end

  defp check_access(conn) do
    # Allow internal service authentication via internal_api_key
    if internal_key_valid?(conn) do
      conn
    else
      # Block direct browser navigation (sec-fetch-dest: document / sec-fetch-mode: navigate)
      if direct_browser_navigation?(conn) do
        forbidden(conn, "Direct browser access is blocked")
      else
        # Allow test environment bypass if in :test and no invalid origin is explicitly provided
        if test_env?() and not invalid_origin_supplied?(conn) do
          conn
        else
          # Validate Origin or Referer header against allowed origins
          origin = get_first_req_header(conn, "origin")
          referer = get_first_req_header(conn, "referer")

          cond do
            origin != nil and allowed_origin?(origin) ->
              conn

            referer != nil and allowed_origin?(referer) ->
              conn

            true ->
              forbidden(conn, "Direct access or untrusted origin is blocked")
          end
        end
      end
    end
  end

  defp internal_key_valid?(conn) do
    case Application.get_env(:presence, :internal_api_key) do
      key when is_binary(key) and key != "" ->
        auth_header = get_first_req_header(conn, "authorization")
        internal_key = get_first_req_header(conn, "x-internal-key")

        auth_header == "Bearer " <> key or internal_key == key

      _ ->
        false
    end
  end

  defp test_env? do
    try do
      Mix.env() == :test
    rescue
      _ -> Application.get_env(:presence, :env) == :test
    end
  end

  defp invalid_origin_supplied?(conn) do
    case get_first_req_header(conn, "origin") do
      nil -> false
      origin -> not allowed_origin?(origin)
    end
  end

  defp direct_browser_navigation?(conn) do
    sec_dest = get_first_req_header(conn, "sec-fetch-dest")
    sec_mode = get_first_req_header(conn, "sec-fetch-mode")

    sec_dest in ["document", "nested-document"] or sec_mode == "navigate"
  end

  def allowed_origin?(origin_or_referer) when is_binary(origin_or_referer) do
    uri = URI.parse(origin_or_referer)
    allowed_list = get_allowed_origins()

    raw_origin =
      case uri.scheme do
        scheme when scheme in ["http", "https"] ->
          port_str = if uri.port && uri.port not in [80, 443], do: ":#{uri.port}", else: ""
          "#{scheme}://#{uri.host}#{port_str}"

        "tauri" ->
          "tauri://#{uri.host}"

        _ ->
          origin_or_referer
      end

    if raw_origin in allowed_list do
      true
    else
      # Check loopback host aliases
      Enum.any?(allowed_list, fn allowed ->
        allowed_uri = URI.parse(allowed)

        allowed_uri.scheme == uri.scheme and
          allowed_uri.port == uri.port and
          allowed_uri.host in @loopback_hosts and
          uri.host in @loopback_hosts
      end)
    end
  end

  def allowed_origin?(_), do: false

  defp get_allowed_origins do
    case Application.get_env(:presence, :allowed_origins) do
      origins when is_list(origins) ->
        origins ++ @default_allowed_origins

      origins when is_binary(origins) and origins != "" ->
        (String.split(origins, ",") |> Enum.map(&String.trim/1)) ++ @default_allowed_origins

      _ ->
        @default_allowed_origins
    end
  end

  defp get_first_req_header(conn, header) do
    case get_req_header(conn, header) do
      [val | _] -> val
      [] -> nil
    end
  end

  defp forbidden(conn, message) do
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(403, Jason.encode!(%{error: "forbidden", message: message}))
    |> halt()
  end
end
